node/actors/bft/src/testonly/twins/partition.rs

```diff
@@ -0,0 +1,115 @@
+use std::cmp::min;
+
+/// A group of nodes that can talk to each other.
+pub type Partition<'a, T> = Vec<&'a T>;
+/// A division of nodes into disjunct partitions, with no communication between different groups.
+pub type Partitioning<'a, T> = Vec<Partition<'a, T>>;
+
+/// Generate all possible partitioning of `items` into `num_partitions` groups.
+///
+/// The idea is to fill out a table such as this:
+/// ```text
+///   P1 P2 P3  
+/// A  *
+/// B  *
+/// C     *
+/// D       *
+/// ```
+/// where the rows are the items and the columns the partitions.
+///
+/// One goal is to minimise the redundant combinations. For example if we want to
+/// partition `[A, B, C]` into two groups, we want `[{A, B}, {C}]` to appear in the
+/// results, but not `[{C}, {A, B}]`, or `[{B, A}, {C}]` as they are the same, but
+/// we do want `[{A, C}, {B}]` because they have different labels.
+pub fn partitions<T>(items: &[T], num_partitions: usize) -> Vec<Partitioning<T>> {
+    Partitioner::generate(items.iter().collect(), num_partitions)
+}
+
+/// Recursive partition generator.
+struct Partitioner<'a, T> {
+    items: Partition<'a, T>,
+    num_items: usize,
+    num_partitions: usize,
+    // All collected complete partitionings
+    output: Vec<Partitioning<'a, T>>,
+    // Partially complete partitioning currently being built
+    acc: Partitioning<'a, T>,
+}
+
+impl<'a, T> Partitioner<'a, T> {
+    /// Generate all possible partitioning.
+    fn generate(items: Partition<'a, T>, num_partitions: usize) -> Vec<Partitioning<'a, T>> {
+        if num_partitions == 0 || items.len() < num_partitions {
+            // Impossible to partition.
+            return Vec::new();
+        } else if num_partitions == items.len() {
+            // Each items stands alone.
```

Comment:
I meant `return` instead of `else`. It doesn't make sense to have both.

Reply:
Sorry, I don't understand. The `else` has been removed.

Reply:
? `} else if num_partitions == items.len() {`

Reply:
Yes? `else if` with `return`, not `else` 🤔 

What used to be the `else` starts at `// Create empty`

Reply:
oh, I meant removing both `else` and replacing `else if` with `if` as the non-happy path exits anyway.